src/data_storage/validators.py

```diff
@@ -0,0 +1,59 @@
+from itertools import chain
+
+from utilities.exceptions import (
+    ExistsCircleValidationError,
+    RootDoesNotExistValidationError,
+    UnconnectedNodeValidationError,
+    UnrelatedReferenceValidationError,
+)
+
+
+def deep_first_search(graph, current_vertex, color=None):
+    if color is None:
+        color = {}
+
+    # If we have reference to None or node without options it means branch has end
+    if current_vertex is None or not graph[current_vertex]:
+        return False
+    color[current_vertex] = "grey"
```

Comment:
Color dict story only "grey" value, so mb we can use boolean instead of string?
And does `visited` name make sense?